umbraco/Umbraco-CMS
# Add extension model for new backoffice

### Prerequisites


### Description

This PR adds the serverside extension model handling for the new backoffice.

The current `package.manifest` format has been replaced with a much simpler format, since the server no longer cares about the actual extensions within the packages. The new extension files are named `umbraco-package.json` and consist of:

- A name
- A version number
- Telemetry allowance (true/false)
- Extensions (which is an array of unknown types)

The `umbraco-package.json` files will still reside under `App_Plugins` - as are the accompanying package resources (scripts etc.) expected to.

We've discussed the naming of the new package format, and the consensus is that it should take over the current `PackageManifest` naming. This means that the current `PackageManifest` and all (almost all) handling thereof has been renamed to `Legacy`, with the intent of deleting it all before V13 ships. Thus... this PR is rather huge, due to all the renaming.

### Testing this PR

1. Create some `umbraco-package.json` files in various folders under `App_Plugins` (see a sample file below).
    - Make sure to create some with and some without telemetry allowed.
2. Verify that all files are detected and returned from the endpoint: `/umbraco/management/api/v1.0/package/manifest`.
3. Create some invalid `umbraco-package.json` files under `App_Plugins` (i.e. non-JSON files). 
4. Restart the site and verify that these invalid files are ignored by the endpoint.
5. Attach a debugger to the site and verify that the `ReportSiteTask` reports the valid packages that have telemetry enabled.
    - It is probably a good idea to set the delay to something less than five minutes, unless you're a big fan of waiting: 
    
![image](https://user-images.githubusercontent.com/7405322/220943654-26d7ad9c-175a-4564-9bc7-f5af8d64e938.png)

### Quirky note

The `ISystemTextJsonSerializer` from the management API is no longer necessary, because we have introduced the `ContextualizedSerializer` in a previous PR 😄
